src/react/components.js

```diff
@@ -153,6 +154,47 @@ function deeplyApplyInstancePrototypeProperties(
   }
 }
 
+export function createClassInstanceForFirstRenderOnly(
+  realm: Realm,
+  componentType: ECMAScriptSourceFunctionValue,
+  props: ObjectValue | AbstractValue,
+  context: ObjectValue | AbstractValue
+): ObjectValue {
+  let instance = Construct(realm, componentType, [props, context]);
+  invariant(instance instanceof ObjectValue);
+  instance.refuseSerialization = true;
+  // assign props
+  Properties.Set(realm, instance, "props", props, true);
+  // assign context
+  Properties.Set(realm, instance, "context", context, true);
+  // assign a mocked setState
+  let setState = new NativeFunctionValue(realm, undefined, `setState`, 1, (_context, [state, callback]) => {
```

Comment:
Instead of mutating setState directly, we should change the `.updater` which what React normally does and includes replaceState and that stuff too.

This can otherwise lead to weird effects due to this being an own property and overriding inheritance chains when people override setState which is on the base class.

Reply:
Let's follow up in a separate PR to reuse `React.Component` logic from the `react` package which uses the updater in a separate function.